derp/derphttp/derphttp_client.go

```diff
@@ -555,24 +691,46 @@ func (c *Client) connect(ctx context.Context, caller string) (client *derp.Clien
 	}
 	if c.preferred {
 		if err := derpClient.NotePreferred(true); err != nil {
-			go httpConn.Close()
+			go connData.httpConn.Close()
 			return nil, 0, err
 		}
 	}
 
 	if c.WatchConnectionChanges {
 		if err := derpClient.WatchConnectionChanges(); err != nil {
-			go httpConn.Close()
+			go connData.httpConn.Close()
 			return nil, 0, err
 		}
 	}
 
 	c.serverPubKey = derpClient.ServerPublicKey()
 	c.client = derpClient
 	c.netConn = tcpConn
-	c.tlsState = tlsState
+	c.tlsState = &connData.tlsState
 	c.connGen++
 
+	if !idealNodeInRegion && !c.attemptingIdealConnection && reg != nil {
+		c.attemptingIdealConnection = true
+
+		go func() {
```

Comment:
we try to avoid long-lived goroutines that aren't named methods/funcs, as it makes debugging goroutine dumps harder.  can this be moved to a method on c?